# 01. Intro homework

Imports

In [3]:
import numpy as np
import pandas as pd

Getting the data

In [4]:
df = pd.read_csv('https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/main/cohorts/2026/data/car_fuel_efficiency_2026.csv')
df

,model_year,origin,fuel_type,drivetrain,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
0,2006,Europe,Gasoline,Front-wheel drive,4,2180,6,243.0,3870,NaN,31.9
1,2008,Europe,Diesel,Front-wheel drive,4,2390,6,272.0,4210,NaN,31.3
2,1996,Asia,Gasoline,Front-wheel drive,5,2320,6,267.0,4240,17.3,27.5
3,1989,Europe,Gasoline,Front-wheel drive,4,2130,6,258.0,4490,18.7,28.5
4,1994,USA,Diesel,Front-wheel drive,3,2580,7,304.0,4510,17.5,31.0
...,...,...,...,...,...,...,...,...,...,...,...
9995,2005,USA,Gasoline,All-wheel drive,4,2470,6,271.0,4690,18.9,27.4
9996,1993,Asia,Gasoline,All-wheel drive,2,2300,6,244.0,4370,19.0,30.0
9997,2014,USA,Gasoline,All-wheel drive,3,2480,6,267.0,4590,18.4,28.1
9998,2004,USA,Hybrid,Rear-wheel drive,5,2180,6,270.0,4450,20.9,32.2


## Q1. Pandas version

In [5]:
pd.__version__

'3.0.6'

Answer: 3.0.6

## Q2. Records count

It сan be found by printing df or df.tail() and looking at the index at the end, then adding 1 to it (also, when you print df, the number of rows and columns is shown at the bottom, here it will be 10000 rows × 11 columns)

In [6]:
df

,model_year,origin,fuel_type,drivetrain,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
0,2006,Europe,Gasoline,Front-wheel drive,4,2180,6,243.0,3870,NaN,31.9
1,2008,Europe,Diesel,Front-wheel drive,4,2390,6,272.0,4210,NaN,31.3
2,1996,Asia,Gasoline,Front-wheel drive,5,2320,6,267.0,4240,17.3,27.5
3,1989,Europe,Gasoline,Front-wheel drive,4,2130,6,258.0,4490,18.7,28.5
4,1994,USA,Diesel,Front-wheel drive,3,2580,7,304.0,4510,17.5,31.0
...,...,...,...,...,...,...,...,...,...,...,...
9995,2005,USA,Gasoline,All-wheel drive,4,2470,6,271.0,4690,18.9,27.4
9996,1993,Asia,Gasoline,All-wheel drive,2,2300,6,244.0,4370,19.0,30.0
9997,2014,USA,Gasoline,All-wheel drive,3,2480,6,267.0,4590,18.4,28.1
9998,2004,USA,Hybrid,Rear-wheel drive,5,2180,6,270.0,4450,20.9,32.2


In [7]:
df.tail()

,model_year,origin,fuel_type,drivetrain,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
9995,2005,USA,Gasoline,All-wheel drive,4,2470,6,271.0,4690,18.9,27.4
9996,1993,Asia,Gasoline,All-wheel drive,2,2300,6,244.0,4370,19.0,30.0
9997,2014,USA,Gasoline,All-wheel drive,3,2480,6,267.0,4590,18.4,28.1
9998,2004,USA,Hybrid,Rear-wheel drive,5,2180,6,270.0,4450,20.9,32.2
9999,2011,USA,Diesel,Front-wheel drive,4,2380,6,295.0,4370,17.6,30.2


Or by looking at df.shape[0], as in numpy, which gives the number of rows of the matrix (dataframe)

In [8]:
df.shape[0]

10000

Answer: 10000

## Q3. Fuel types

Can be found using nunique(), which counts the number of unique values in a column

In [9]:
df.fuel_type.nunique()

3

Answer: 3

## Q4. Missing values

Can be found using df.isnull().sum(), as shown in the course video. It shows the number of missing values in each column

In [10]:
df.isnull().sum()

model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors                0
engine_displacement      0
num_cylinders            0
horsepower             877
vehicle_weight           0
acceleration           264
fuel_efficiency_mpg      0
dtype: int64

There are 2 such columns. It can also be found like this

In [11]:
(df.isnull().sum() > 0).sum()

np.int64(2)

df.isnull().sum() counts missing values in each column, > 0 checks which columns have at least one, and the last .sum() counts those columns

Answer: 2

## Q5. Max fuel efficiency

Can be found by filtering cars from Asia first and then taking the maximum of fuel_efficiency_mpg column

In [12]:
df[df['origin'] == 'Asia'].fuel_efficiency_mpg.max()

np.float64(41.2)

Answer: 41.2

## Q6. Median value of horsepower

Median of horsepower

In [13]:
median_before = df.horsepower.median()
median_before

np.float64(254.0)

Most frequent value

In [14]:
mode_horsepower = df.horsepower.mode()[0]
mode_horsepower

np.float64(252.0)

In [15]:
median_after = df.horsepower.fillna(mode_horsepower).median()
median_after

np.float64(252.0)

The median decreased from 254.0 to 252.0.

Answer: Yes, it decreased

## Q7. Sum of weights

Steps 1-4

In [16]:
X = df[df['origin'] == 'Asia'][['vehicle_weight', 'model_year']].head(7).values
X

array([[4240, 1996],
       [4310, 1992],
       [4230, 1997],
       [4420, 1978],
       [3950, 2007],
       [4750, 2021],
       [4450, 2019]])

Step 5

In [17]:
XTX = X.T.dot(X) # or X.T @ X
XTX

array([[131950500,  60750580],
       [ 60750580,  28041424]])

Step 6

In [18]:
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 2.96830537e-06, -6.43071025e-06],
       [-6.43071025e-06,  1.39675281e-05]])

Step 7

In [19]:
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
y

array([1100, 1300,  800,  900, 1000, 1100, 1200])

Steps 8-9

In [20]:
w = XTX_inv.dot(X.T).dot(y).sum()
w

np.float64(0.36919696904927707)

Answer: 0.369